docs/runtime-contract.md

```diff
@@ -0,0 +1,445 @@
+implementations MAY strip certain HTTP headers for security or other reasons;
+such implementations SHOULD document the set of stripped headers. Because the
+full set of HTTP headers is constantly evolving, it is RECOMMENDED that
+platforms which strip headers define a common prefix which covers all headers
+removed by the platform.
+
+Also, the following proxy-specific request headers MUST be set, in addition to
+the base set of HTTP/1.1 headers (e.g. `Host:`)
+
+- `Forwarded` - As specified by [RFC 7239](https://tools.ietf.org/html/rfc7239).
+  Additionally, the following legacy headers SHOULD be set for compatibility
+  with client software:
+  - `X-Forwarded-For`
+  - `X-Forwarded-Proto`
+
+In addition, the following headers SHOULD be set to enable tracing and
+observability features:
+
+- Trace headers - Platform providers SHOULD provide and document headers needed
+  to propagate trace contexts, in the absence of w3c standardization.
+
+Operators and platform providers MAY provide additional headers to provide
+environment specific information.
+
+#### Meta Requests
+
+The
+`[core.v1.Container](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#container-v1-core)`
+object allows specifying both a
+[`readinessProbe`](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-readiness-probes)
+and a
+[`livenessProbe`](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-a-liveness-http-request).
+If not provided, container startup and listening on the declared HTTP socket is
+considered sufficient to declare the container "ready" and "live" (see the probe
+definition below). Developers MUST use the `httpGet` or `tcpSocket` probe method
+on the inbound container port for readiness and liveness checks; platform
+providers SHOULD disallow other probe methods.
+
+Because serverless platforms automatically scale instances based on inbound
+requests, the following defaults SHOULD be applied by the platform provider if
+not set by the developer. The probes are intended to be trivially supportable by
+naive conforming containers while preventing interference with developer code.
+These settings apply to both `livenessProbe` and `readinessProbe`:
+
+- `tcpSocket` set to the container's port
+- `initialDelaySeconds` set to 0
+- `periodSeconds` set to platform-specific value
+
+In order to enable scaling in response to load, developers SHOULD NOT set the
+`initialDelaySeconds` to a value greater than 0, and should aim to minimize
+container startup time (aka cold start time).
+
+##### Deployment probe
+
+On the initial deployment, platform providers SHOULD start an instance of the
+container to validate that the container is valid and will become ready. This
+startup SHOULD occur even if the container would not serve any user requests. If
+a container cannot satisfy the `livenessProbe` and `readinessProbe` during
```

Comment:
> With that said, it's definitely intended as a "SHOULD", and it would be okay to for an operator not to do this and have developers discover misconfigurations late if they slip through CI/CD.

Right - and we'd like for this to be available to our operators w/o forking the Kn code base.